### DISCUSSION TITLE
Document Upload (In Progress)

### DISCUSSION THREAD

**Discussion by cbutler-captech** [microsoft/sample-app-aoai-chatGPT#713] [Ideas]
All, I have a team internally working on a document upload feature (chat with your document).  The goal is secure and isolated chat with oneoff documents.  We are building this internally but have plans to push it back as a pull request when complete.

Once complete you will be able to:

Upload a document
Chat with it - ensuring that noone else can access it.  Document will only be available to the current conversation (or user, depending on configuration)
Chat with any documents marked as global (configurable)

Tech Stack:
Azure AI Search
Azure OpenAI/Vector Embeddings
Azure Storage Blobs

![image](https://github.com/microsoft/sample-app-aoai-chatGPT/assets/149815129/3acd69bc-a8a0-4cc1-a179-7fe7a5a8bce0)

Please add your thoughts...

**Comment by ianlcassidy** [comment]
This would be a great addition to this repo!

**Reply by cbutler-captech** [reply]
Update on this.  This is going into internal testing this week.  We have successfully uploaded documents, chatted with them and restricted them by either conversation or user such that they are private and noone else can see them and the chat just returns a generic result on other conversations.  Working on some front end items and edge cases.

Other features:
Delete blobs on conversation delete (configurable)
Indexer cloning to increase concurrency
Folder configuration
Global document configuration by tag

**Comment by ianlcassidy** [comment]
@cbutler-captech - thanks for the update. do you know if this functionality will allow you to upload a document and compare it to an existing AI Search index? for example, you upload a document and then ask the bot to make any suggestions or point out gaps based on similar documents that have already been indexed? i have not seen a tool for doing this yet, but am interested in this functionality.

**Reply by cbutler-captech** [reply]
If those documents are accessible (forgot we also have global document configuration), it is likely that it can do this.

**Reply by cbutler-captech** [reply]
Note we are testing with gpt-3.5 just to keep costs low for now.  A great addition to this repo is model selection by conversation.  I will add a new thread for that.

**Reply by ianlcassidy** [reply]
got it, that all makes sense.

**Comment by ianlcassidy** [comment]
hey @cbutler-captech - any updates on this?

**Reply by cbutler-captech** [reply]
This is in our internal testing.  Adding a few other things to finish it up like upload limits.  Also currently loses track of uploads/vector index status if you switch to a new conversation, looking into picking these back up on conversation load.

**Reply by ianlcassidy** [reply]
amazing, thanks!

**Reply by cbutler-captech** [reply]
Slow going here with alot of edge cases and security reviews internally.  Still need to work on documentation as its quite a large and complex feature.

**Reply by ianlcassidy** [reply]
i am not surprised. thanks for the update. is it worth putting it behind some kind of feature flag in the env vars and have the community test and help debug the functionality?

**Reply by cbutler-captech** [reply]
It would be disabled by default, as you would need to configure some storage and indexing before it would function.  Document retention and expiration is another discussion item.

**Reply by eshwarhs** [reply]
Can you put up a WIP open PR so that we can see the code and test it out as well?

**Reply by cbutler-captech** [reply]
I have a first pass at the documentation done, once we get the rework of the index polling mechanism completed I will push it out there

**Reply by jasredyy5** [reply]
Hey @cbutler-captech, do you have an estimate on when this can become a PR? Thank you!

**Reply by cbutler-captech** [reply]
Still jumping through internal hoops.  Trying to push this forward.

**Reply by jasredyy5** [reply]
Thank You. Looking forward to this!

**Comment by cbutler-captech** [comment]
Thank you for your patience.  We are new to the open source world and I had to make sure I had all my ducks in a row.  Setup is kindof tricky.  See the readme_docupload.md and each out if you have any issues.

https://github.com/microsoft/sample-app-aoai-chatGPT/pull/862

**Reply by jeannot-prvlimburg** [reply]
Excellent work! I have a proposal for a small adaptation. It would be beneficial to add an option to load the file content directly into the chat context without using blob storage. This would simplify the process by eliminating the need to save the file or make it accessible to others. It would enhance the functionality and make it even more user-friendly. What do you think?

EDIT: To clarify further, we want to maintain control over which documents are stored in the VectorDB, but we would like to give users the ability to upload documents directly into the context of a chat.

**Reply by jasredyy5** [reply]
> Thank you for your patience. We are new to the open source world and I had to make sure I had all my ducks in a row. Setup is kindof tricky. See the readme_docupload.md and each out if you have any issues.
> 
> #862

We are trying to set this up and test the functionality locally but the indexer is failing to run with the below error. 
"Web Api response status: 'Unauthorized', Web Api response details: '{"error":{"code":"PermissionDenied","message": "Principal does not have access to API/Operation."}}'"

Do we need any special API permissions or custom roles for this to work? Thanks in advance!

**Reply by cbutler-captech** [reply]
I may need to update the security section to be more helpful,  Basically, all managed servies need rights to the other services.

- Blob storage
  - Grant Reader and Data Access to both your OpenAI and AISearch managed identities
  - Grant Storage and Blob Data Reader to you r OpenAI and AISearch managed identities
- OpenAI
  - Grant Azure AI Developer to your AI Search Managed identity (may be need to examine for a lesser privilige)

**Reply by cbutler-captech** [reply]
> Excellent work! I have a proposal for a small adaptation. It would be beneficial to add an option to load the file content directly into the chat context without using blob storage. This would simplify the process by eliminating the need to save the file or make it accessible to others. It would enhance the functionality and make it even more user-friendly. What do you think?
> 
> EDIT: To clarify further, we want to maintain control over which documents are stored in the VectorDB, but we would like to give users the ability to upload documents directly into the context of a chat.

Technically speaking the blobs are unnecessary after getting into the vector search.  If you can figure out the code to replicate what AI search is doing to create the chunks and just push them directly in there with the conversation and userids, that would definately be better.  I looked into it for a while but was going in circles and all docs were telling me to use AI Search.  I know I can make the vectors and chunks but AI Search does a little more and we decided to let it be.

**Reply by jasredyy5** [reply]
> I may need to update the security section to be more helpful, Basically, all managed servies need rights to the other services.
> 
> * Blob storage
>   
>   * Grant Reader and Data Access to both your OpenAI and AISearch managed identities
>   * Grant Storage and Blob Data Reader to you r OpenAI and AISearch managed identities
> * OpenAI
>   
>   * Grant Azure AI Developer to your AI Search Managed identity (may be need to examine for a lesser privilige)

This is helpful. Thank You!

**Reply by eshwarhs** [reply]
@cbutler-captech I set up the code locally and I observed this behaviour - 
After a file is uploaded in a conversation and a new conversation is started using the '+', any new file uploaded, is appending it to the previous conversation. This is probably due to the callback function "handleUploadSuccess" and "handleConversationIdUpdate" along with the SET_ACTIVE_CONVERSATION_ID event. Can you confirm if this behavior is consistent on your code as well?

**Reply by cbutler-captech** [reply]
I will look into this one.  My front end expert is no longer available but I may be able to get it.

**Comment by patelha-ccm** [comment]
Hey @cbutler-captech , any update on this feature? When can we expect it to roll out. I am really excited of its use-case specifically within our team :)

**Reply by cbutler-captech** [reply]
We had some more front end improvements going but I think this will be the PR as is.  I will update and take it out of WIP.

**Comment by stbere** [comment]
Thank you! This looks pretty good, can't wait to get this integrated into my app.

**Comment by ukiguy** [comment]
Looking forward to using this feature. Do we have any ETA on this feature release? Thank you for the hard work.

**Comment by misslapina123** [comment]
What is the status of this feature?

**Comment by 2gad** [comment]
Not sure who this is helpful for but there is this other repo (managed by Azure Samples) https://github.com/Azure-Samples/azure-search-openai-demo that has a similar chat app but has a file upload function

I am planning to try to integrate this into what I have forked from sample-app-aoai-chatGPT

**Reply by stbere** [reply]
So I use that repo and this one as well. This repo we are in is tied to the ai.azure.com studio, which I like. but there are stark differences I'm seeing, and one is that the citation for this particular repo is lacking, whereas the link you shared has a beautiful PDF web viewer for the citation panel.

**Comment by cbutler-captech** [comment]
My apologies, my team has disbanded and I no longer have the front end resources to dedicate to this feature.  I hope someone can take this over and get any issues resolved and get it into the main branch as we are evaluating other front ends and the potential to put effort into custom Copilots.

**Reply by davidgristwood** [reply]
That's a real shame. Are you saying this feature is effectively dead unless someone picks up the work and does a pull request?

**Comment by davidgristwood** [comment]
Copilot been getting a lot of attention as a UI and orchestrator - I'd appreciate any insights into whether it has a role in this scenario, or is this a prioritisation on copilot?

**Comment by jeannot-prvlimburg** [comment]
You could switch to [azure-search-openai-demo](https://github.com/Azure-Samples/azure-search-openai-demo), which already has it implented.

**Reply by stbere** [reply]
When did they add this? I've been using both. I do like how the openai-demo now has voice integration, but haven't implemented it into my product yet.
